games.d/wator.d/wator.6
.V= : wator.6 84/12/18 $
.TH WATOR 6 "\*(V)" "Local" "\*(O)"
.SH NAME
wator  \-  A computer recreation from Scientific American
.SH SYNOPSIS
.B wator
[
.B \-hs
] [
.B #sharks sbreed #fish fbreed starve
]
.SH DESCRIPTION
.ad b
This program manipulates an ocean inhabited by sharks and fish.  Both swim and
breed.  The sharks eat fish, and starve if they don't eat for a while.  The
ocean is displayed as a grid of 23 by 80 cells.  If you move up and off the
screen you "wrap around" to the bottom.  The same action occurs for the sides.
Empty cells are blank, sharks are displayed as "O", fish as ".".  Time
progresses as cycles called "chronons".
.PP
Each chronon, the program examines each fish in sequence.  A fish will choose
one adjacent empty cell at random and move to it.  If sufficient chronons
have passed to cause the fish to breed, a new fish will be placed at the
location that this fish has just vacated.  Both fish would then have their
ages set to 0.  If no empty location exists, fish will not be able to move
or breed, though the age will increase.
.PP
Each chronon sharks will be examined in a manner similar to that for the
fish.  The sharks will first examine all adjacent cells to see if any contain
fish.  An adjacent cell containing a fish would be chosen at random and
eaten, with the shark moving to the location which had held the fish.
Should no fish be found in adjacent cells the shark would just choose at
random an adjacent empty cell.  Sharks breed in the same manner as fish,
though the number of chronons between breeding is independent of the time
required for fish to breed.
.PP
Sharks starve if they do not eat for some number of chronons.
Fish only decline by being eaten.  The definition of
.B adjacent
is a compile time option.  It may refer to the four cells on the sides
of a cell, or may refer to all eight cells surrounding another cell.
.PP
Arguments may be supplied like so:
.TP
.B \-h
print a help message
.TP
.B \-s
generate statistics for each chronon in a file
.B wator.stat
.PP
A series of numbers may be given to specify: the initial number of sharks,
the number of chronons required for sharks to breed, the initial number of
fish, the number of chronons required for fish to breed, and then number of
chronons required for sharks to starve.  With no input values some
reasonable starting conditions are chosen at random.
.PP
Information shown on the bottom line is:
generation, #sharks, #fish, when sharks and fish breed, and when sharks
starve.
.SH FILES
/etc/termcap
.SH AUTHOR
Dale Mosby
.SH SEE ALSO
December 1984 Scientific American, Vol 251 Number 6, Page 14
Article "Computer Recreations" by A. K. Dewdney
.SH BUGS
Very little time went into the argument parsing, and almost all testing
and use of the program has been with no arguments, just letting the
program choose values.
.PP
The ability to redraw the screen after the job is stopped and then restarted
was a quick addition and sometimes causes a partial cursor sequence
to be displayed.


games.d/wator.d/wator.c
/*
 * WATOR -- Sharks and fish wage an ecological war on the
 *          toroidal planet Wa-Tor.
 *
 * ------------------------------------------------------------------
 * A computer recreation.  See the December 1984 Scientific American.
 * Volume 251, Number 6, page 14.
 * ------------------------------------------------------------------
 * This program written by Dale Mosby of Sequent Computer Systems, Inc.
 * 14360 NW Science Park Drive,  Portland, Oregon 97229 (503) 626-5700.
 * Permission for not for profit use and distribution is hereby granted
 * providing this acknowledgment is included.  No warranties implied!
 * ------------------------------------------------------------------
 *
 * wator [-hs] #sharks sbreed #fish fbreed starve
 *
 *       -h   help info
 *       -s    generate statistic log file wator.stat
 *
 * See "hlp_msg" at end of this file for information on parameters.  This
 * program displays a flat map of the world.  If you move off one edge you
 * come back on the other.  Sharks and fish are randomly placed on the world.
 * Time proceeds in cycles with fish moving randomly to an adjacent unoccupied
 * cell.  Fish breed after some number of cycles, creating more fish.
 * Each cycle the sharks will choose one adjacent fish and eat it.  Sharks
 * also breed after some number of cycles (not necessarily the same number
 * as for fish).  If sharks find no adjacent fish they move randomly as do
 * fish.  (Read the magazine article for a good explanation).
 */

#include <stdio.h>
#include "wator.h"
#ifndef SYS5
#include <signal.h>
#endif

main(argc, argv)
	int argc;
	char *argv[];
{
#ifndef SYS5
	struct sigvec vec, ovec;
	int sc();

	vec.sv_handler = sc;
	vec.sv_mask = 0;
	vec.sv_onstack = 0;
	if (sigvec(SIGCONT, &vec, &ovec)) {
		printf("Sigvec(SIGCONT) fails.\n");
		exit(1);
	}
#endif
	init();
	parse_args(argc, argv);
	screen_control();
	init_free();
	init_screen();
	seed_fish();
	seed_shark();
	draw_screen();
	put_stat();

	do {
		generation++;
		fish_cycle();
		shark_cycle();
		put_stat();
	} while ( (num_shark && num_fish) &&
		  (num_shark+num_fish < wide * lines-1) );

	done_msg();
}

#ifndef SYS5
/*
 * After we get the continue signal, assume screen is messed up so redraw it.
 * If we get clobbered in the middle of some IO there could be a botched cursor
 * control sequence going out.  A second redraw in the middle of the main
 * "while" loop would pretty much guarantee that things were fine.
 */
sc()
{
	draw_screen();
	put_stat();
}
#endif

done_msg()
{
	int i;

	movexy(wide-12, lines-1);
	printf("Hit char");
	i = getchar();
}

char *hlp_msg[]= {
"\n",
"WATOR - sharks and fish wage an ecological war on the toroidal planet Wa-Tor.\n\n",
"Based on the 'Computer Recreations' column from the December 1984 Scientific\n",
"American, Vol 251, Number 6, page 14.  By A. K. Dewdney\n\n",
"With no arguments this program chooses random values and begins.\n",
"Sharks display as 'O', and fish as '.'  Arguments:",
"\n\nwator [-hls] #sharks #fish sbreed fbreed starve\n",
" -h -- this help message\n",
" -s -- generate statistics in a file (default /tmp/wator.stat)\n\n",
"After these options specify 5 numbers:\n",
"   #sharks - initial shark population\n",
"   #fish   - innitial fish population\n",
"   sbreed  - how many generations before sharks breed\n",
"   fbreed  - how many generations before fish breed\n",
"   starve  - how many generations without eating before sharks starve\n\n",
"Example:   'wator 50 250 6 4 5'\n\n",
"Final line gives statistics: generation number, number of sharks, number\n",
"of fish, how often sharks/fish breed, and when sharks starve.\n",
0 };

usage()
{
	char **p;

	p = hlp_msg;
	while (*p) {
		printf(*p);
		p++;
	}
}
